x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractBlock.java

```diff
@@ -101,4 +100,54 @@ public BlockFactory blockFactory() {
     public boolean isReleased() {
         return released;
     }
+
+    @Override
+    public final void incRef() {
+        if (references <= 0) {
+            throw new IllegalStateException("can't increase refCount on already released block [" + this + "]");
+        }
+        references++;
+    }
+
+    @Override
+    public final boolean tryIncRef() {
+        if (references <= 0) {
+            return false;
+        }
+        references++;
+        return true;
+    }
+
+    @Override
+    public final boolean decRef() {
+        if (references <= 0) {
+            throw new IllegalStateException("can't release already released block [" + this + "]");
+        }
+
+        references--;
+
+        if (references <= 0) {
+            close();
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public final boolean hasReferences() {
+        return references >= 1;
+    }
+
+    @Override
+    public void close() {
+        if (released) {
```

Comment:
While we could enforce a close-once behavior, I prefer allowing multiple-closes to leverage the try-resource.